### ISSUE TITLE
MacOS: LSOpenURLsWithRole() failed with error -10810

### ISSUE THREAD

**Issue by GordonMcKinney** [pkulchenko/ZeroBraneStudio#1018]
Running ZeroBranStudio v1.80 and MacOS 10.15 (Catalina)

```
Applications $ open ZeroBraneStudio.app
LSOpenURLsWithRole() failed with error -10810 for the file /Applications/ZeroBraneStudio.app.
```

**Comment by pkulchenko** [comment]
@GordonMcKinney, thank you for the report; this should be fixed shortly in the upgrade branch. I'll update the ticket when the change is pushed out.

Ref #999.

**Comment by GordonMcKinney** [comment]
@pkulchenko thank you! Should I pull a dev branch build and test?

**Comment by pkulchenko** [comment]
@GordonMcKinney, you can check out branch https://github.com/pkulchenko/ZeroBraneStudio/tree/all-binaries-190b. you don't need to build anything, just run `./zbstudio.sh` after cloning and checking out `all-binaries-190b`.

**Comment by GordonMcKinney** [comment]
@pkulchenko That did the trick! I will be using this over the next couple of days to meet some deadlines, so I really appreciate your quick response. I'll raise a new issue if I see anything awry.

PS. I made a donation! :-)

**Comment by pkulchenko** [comment]
@GordonMcKinney, thank you for the update (and donation)! These changes will be included in the next release. Please update this ticket (or #999) if you run into any issues with the new build.

**Comment by KevBruner** [comment]
Hello! I am also experiencing this issue, but it continues in the all-binaries-190b branch as well. Thanks.

**Comment by pkulchenko** [comment]
@KevBruner, any further details? Are you sure you ran `git checkout all-binaries-190b` after cloning? Did you then launch the IDE as `./zbstudio.sh`? If you still get the error, can you also try `bin/lua src/main.lua` instead?

**Comment by KevBruner** [comment]
I'm trying to track it down now. I'm getting a `Symbol not found: _lua_cpcall` which seems to be coming from HomeBrew mapping all Lua versions to the "installed homebrew" version. So ZBS is trying to load Lua5.1 but seems to be finding a dylib which is actually Lua5.3

**Comment by pkulchenko** [comment]
Yes, this looks like the case of version mismatch (Lua5.3 or Lua5.2 binary trying to load Lua5.1 dynamic library). homebrew shouldn't be involved with ZBS execution, as it comes with its own Lua executables and libraries. Or do you get this error when you try to launch an application *from* the IDE?

Can you try running `bin/lua src/main.lua` from the IDE folder.

**Comment by KevBruner** [comment]
When I run `bin/lua src/main.lua` from the IDE folder then it works! I was surprised to see ZBS attempt to use the libraries from `/usr/local/...` but that's the error I was seeing.

**Comment by pkulchenko** [comment]
@KevBruner, thank you for the update. How did you start the application before? What was the current directory and how were you launching it?

Can you also try `open zbstudio/ZeroBraneStudio.app` from the IDE folder?

**Comment by KevBruner** [comment]
I opened the app by using `./zbstudio.sh` (which causes the error), and `open zbstudio/ZeroBraneStudio.app` also causes the error, but `bin/lua src/main.lua` does not. Seems like the way my system finds dylibs might be somehow funky...

**Comment by pkulchenko** [comment]
Hm, can you do one more thing? open `zbstudio/ZeroBraneStudio.app/Contents/MacOS/ZeroBraneStudio` file (it's a shell script) and change it to the following:

```sh
#!/bin/bash
ZBS_PATH=${0%/*/*}
echo 1 $ZBS_PATH >~/zbs-debug.log
if [ ! -d $ZBS_PATH ]; then ZBS_PATH=${PWD%/*}; fi
echo 2 $ZBS_PATH >>~/zbs-debug.log
echo 3 `pwd` >>~/zbs-debug.log
cd "$ZBS_PATH/ZeroBraneStudio"
echo 4 `pwd` >>~/zbs-debug.log
exec bin/lua src/main.lua zbstudio "$@"
```

Then launch the script using `./zbstudio.sh` and `open zbstudio/ZeroBraneStudio.app` and capture the result in the debug file each time.

The only reason for the error I can think of is that the app is launching from the wrong location, which this test will demonstrate (I hope). I can't think of anything else, given that bin/lua loads its libraries based on `@executive_path` (which you may confirm by running `otool -l bin/lua`.

**Comment by KevBruner** [comment]
Here you go! (Sorry for the delay, had a phone call)

`./zbstudio.sh`
```1 /Users/kbruner/Desktop/ZeroBraneStudio-all-binaries-190b/zbstudio/ZeroBraneStudio.app/Contents
2 /Users/kbruner/Desktop/ZeroBraneStudio-all-binaries-190b/zbstudio/ZeroBraneStudio.app/Contents
3 /
4 /Users/kbruner/Desktop/ZeroBraneStudio-all-binaries-190b/zbstudio/ZeroBraneStudio.app/Contents/ZeroBraneStudio
```

`open zbstudio/ZeroBraneStudio.app`

```1 /Users/kbruner/Desktop/ZeroBraneStudio-all-binaries-190b/zbstudio/ZeroBraneStudio.app/Contents
2 /Users/kbruner/Desktop/ZeroBraneStudio-all-binaries-190b/zbstudio/ZeroBraneStudio.app/Contents
3 /
4 /Users/kbruner/Desktop/ZeroBraneStudio-all-binaries-190b/zbstudio/ZeroBraneStudio.app/Contents/ZeroBraneStudio```

**Comment by pkulchenko** [comment]
This all looks correct; so it still works when you run the same command `bin/lua src/main.lua` from the same folder `/Users/kbruner/Desktop/ZeroBraneStudio-all-binaries-190b/zbstudio/ZeroBraneStudio.app/Contents/ZeroBraneStudio`, right?

You can try removing `exec` from the last line in the shell script, but I don't see why it would make this difference. In fact `exec` was added to make single instance detection to work properly on High Sierra (#813).

**Comment by pkulchenko** [comment]
@KevBruner, one more thing to check. What does `xattr -pl com.apple.quarantine zbstudio` command return from the `ZeroBraneStudio` folder?

**Comment by KevBruner** [comment]
Removing `exec` didn't seem to make any difference. Still only works is I run it as `bin/lua src/main.lua`. Very strange, I'll continue to investigate.

**Comment by pkulchenko** [comment]
I agree, it is indeed strange. Let me know what you can find.

**Comment by KevBruner** [comment]
`xattr: zbstudio: No such xattr: com.apple.quarantine` is what I see from `/Users/kbruner/Desktop/ZeroBraneStudio-all-binaries-190b`. Is theat the `ZeroBraneStudio` folder you referenced? I don't see a folder names `ZeroBranStudio` exactly.

**Comment by pkulchenko** [comment]
Yes, that's the one; thank you. I don't have any other ideas for now. Will check on a couple of things, but nothing jumps at me as being wrong so far.

**Comment by KevBruner** [comment]
Thanks for your help! I'll post more as I investigate further.

**Comment by pkulchenko** [comment]
@KevBruner, I think I got it. You didn't clone and check out the branch; you downloaded the zip archive with the branch content, but since the app is not signed, Catalina doesn't allow you to launch it as the app is from an "untrusted" source. You can do `git clone` for this repo and then `git checkout all-binaries-190b` and I expect it to work. When the next release is packaged, the installation bundle will be signed, so this is not going to be an issue.

There is also a command that removes quarantine when you use zbstudio.sh script, but if you try to open the app first (before launching ./zbstudio.sh), the quarantine status appears to be cached somewhere in macOS, so even after it's removed the app is not going to be launched. The remedy is to re-download the zip file with the branch to a different location (maybe not under Desktop) and then run `./zbstudio.sh`.

(1) use clone of the repo, from which all three options (launching the app, `./zbstudio.sh`, and `bin/lua src/main.lua`) will work
(2) download the zip with the branch and launch `./zbstudio.sh` first, as it will remove the quarantine. Then all three launch options should work again.

**Comment by pkulchenko** [comment]
Actually, one more clarification for those who may run into this. lua application in the IDE loads liblua using @rpath. Since the directories under Desktop and Documents are sandboxed, the load of liblua fails, but since @rpath will search in other locations, it finds the one installed by brew and fails there (as it's for the wrong version of Lua), hence the error you saw.

If you move the folder to some other location (not Desktop or Documents), it should work. I'll update the launch script to check if it's under Desktop/Documents and issue a warning to report this.

**Comment by KevBruner** [comment]
Awesome! Thank you so much!

**Comment by HappyGames** [comment]
when i run with ./zbstudio.sh  there is no window show up , just a menu on top of the screen
<img width="590" alt="sss" src="https://user-images.githubusercontent.com/3297604/72403051-ddd61c80-378b-11ea-8eba-b2b88e2afcf8.png"> @pkulchenko

**Comment by pkulchenko** [comment]
@HappyGames, this issue should be fixed in v1.90.
